taichi/rhi/common/memory_pool.cpp

```diff
@@ -0,0 +1,275 @@
+        TI_P(size);
+        TI_ERROR("Insufficient virtual memory space");
+      }
+      ptr =
+          VirtualAlloc(nullptr, size, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
+      TI_ERROR_IF(ptr == nullptr, "Virtual memory allocation ({} B) failed.",
+                  size);
+#endif
+      TI_ERROR_IF(((uint64_t)ptr) % page_size != 0,
+                  "Allocated address ({:}) is not aligned by page size {}", ptr,
+                  page_size);
+    } else {
+      TI_NOT_IMPLEMENTED;
+    }
+
+    if (raw_memory_chunks_.count(ptr)) {
+      TI_ERROR("Memory address ({:}) is already allocated", ptr);
+    }
+
+    raw_memory_chunks_[ptr] = size;
+    return ptr;
+  }
+
+  void deallocate_raw_memory(void *ptr) override {
+    /*
+      Be aware that this methods is not protected by the mutex.
+
+      deallocate_raw_memory() is designed to be a private method, and
+      should only be called by its Allocators friends.
+
+      The caller ensures that no other thread is accessing the memory pool
+      when calling this method.
+    */
+    if (!raw_memory_chunks_.count(ptr)) {
+      TI_ERROR("Memory address ({:}) is not allocated", ptr);
+    }
+
+    std::size_t size = raw_memory_chunks_[ptr];
+#if defined(TI_PLATFORM_UNIX)
+    if (munmap(ptr, size) != 0)
+#else
+    // https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualfree
+    // According to MS Doc: size must be when using MEM_RELEASE
+    if (!VirtualFree(ptr, 0, MEM_RELEASE))
+#endif
+      TI_ERROR("Failed to free virtual memory ({} B)", size);
+
+    raw_memory_chunks_.erase(ptr);
+  }
+
+  void reset() override {
+    std::lock_guard<std::mutex> _(mut_allocation_);
+    allocator_ = std::unique_ptr<UnifiedAllocator>(new UnifiedAllocator(arch_));
+
+    const auto ptr_map_copied = raw_memory_chunks_;
+    for (auto &ptr : ptr_map_copied) {
+      deallocate_raw_memory(ptr.first);
+    }
+  }
+
+  ~HostMemoryPool() {
+    reset();
+  }
+};
+
+class CudaMemoryPool : public MemoryPool {
+ public:
+  CudaMemoryPool(Arch arch) {
+    arch_ = arch;
+    TI_ASSERT(arch_is_cuda(arch_));
+
+    // TODO(zhanlue): replace UnifiedAllocator with CachingAllocator
+    allocator_ = std::unique_ptr<UnifiedAllocator>(new UnifiedAllocator(arch));
+  }
+
+  void *allocate(std::size_t size,
+                 std::size_t alignment,
+                 bool exclusive,
+                 bool managed) override {
+    std::lock_guard<std::mutex> _(mut_allocation_);
+
+    // TODO(zhanlue): Replace UnifiedAllocator with Caching Allocator
+    // Here we reuse the UnifiedAllocator's allocation logic to perform
+    // pre-allocation, but ideally this preallocation logic should be fused into
+    // the Caching Allocator
+    if (!allocator_) {
+      TI_ERROR("Memory pool is already destroyed");
+    }
+
+    void *ret = allocator_->allocate(size, alignment, exclusive, managed);
+    return ret;
+  }
+
+  void release(std::size_t size, void *ptr) override {
```

Comment:
So we release the memory iff `ptr` is the beginning of a chunk? Is there a circumstance that we released the beginning of a chunk but somewhere in the middle of a chunk is still in use?